cpu/stm32f4/periph/eth.c

```
@@ -0,0 +1,540 @@
+    }
+}
+
+void isr_eth(void)
+{
+    volatile unsigned tmp = ETH->DMASR;
+
+    if ((tmp & ETH_DMASR_TS)) {
+        ETH->DMASR = ETH_DMASR_TS | ETH_DMASR_NIS;
+        mutex_unlock(&_tx);
+    }
+
+    if ((tmp & ETH_DMASR_RS)) {
+        ETH->DMASR = ETH_DMASR_RS | ETH_DMASR_NIS;
+        mutex_unlock(&_rx);
+        if (_netdev2) {
+            _netdev2->event_callback(_netdev2, NETDEV2_EVENT_ISR);
+        }
+    }
+
+    /* printf("r:%x\n\n", tmp); */
+
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+void isr_eth_wkup(void)
+{
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+void ETH_DMA_ISR(void)
+{
+    /* clear DMA done flag */
+    int stream = eth_config.dma_stream;
+    dma_base(stream)->IFCR[dma_hl(stream)] = dma_ifc(stream);
+    mutex_unlock(&_dma_sync);
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+static int _send(netdev2_t *netdev, const struct iovec *vector, int count)
+{
+    (void)netdev;
+
+    int ret = 0, len = 0;
+    mutex_lock(&send_lock);
+    for (int i = 0; i < count && ret == 0; i++) {
+        ret = eth_send(vector[i].iov_base, vector[i].iov_len);
+        len += ret;
+    }
+    SDEBUG("_send: %d %d\n", ret, len);
+    mutex_unlock(&send_lock);
+
+#ifdef MODULE_NETSTATS_L2
+    netdev->stats.tx_bytes += len;
+#endif
+
+    return len;
+}
+
+static int _recv(netdev2_t *netdev, char *buf, int len, void *info)
+{
+    (void)info;
+    (void)netdev;
+
+    int ret = _try_receive(buf, len, 1);
+#ifdef MODULE_NETSTATS_L2
+    if (buf) {
+        netdev->stats.rx_count++;
+        netdev2->stats.rx_bytes += len;
+    }
+#endif
+
+    SDEBUG("_recev: %d\n", ret);
+
+    return ret;
+}
+
+static int _get(netdev2_t *dev, netopt_t opt, void *value, size_t max_len)
+{
+    int res = -1;
+
+    switch (opt) {
+        case NETOPT_ADDRESS:
+            if (max_len >= ETHERNET_ADDR_LEN) {
```

This would fail silently. Until now we used asserts for checking the length.